imap_processing/lo/l0/science_direct_events.py

```diff
@@ -0,0 +1,202 @@
+
+    The Science Direct Events class handles the parsing and
+    decompression of L0 to L1A data.
+
+    Attributes
+    ----------
+    SHCOARSE : int
+        Spacecraft time.
+    COUNT: int
+        Number of direct events.
+    CHKSUM: int
+        Checksum for the packet.
+    DATA: bitstring.Bits
+        Compressed TOF Direct Event time tagged data.
+    TOF0: int
+        Time of Flight 0 value for direct event.
+    TOF1: int
+        Time of Flight 1 value for direct event.
+    TOF2: int
+        Time of Flight 2 value for direct event.
+    TOF3: int
+        Time of Flight 3 value for direct event.
+    TIME: int
+        time tag for the direct event
+    ENERGY: int
+        energy of the direct event ENA.
+    POS: int
+        Stop position for the direct event. There are 4 quadrants
+        on the at the stop position.
+    CKSM: int
+        This is checksum defined relative to the TOFs
+        condition for golden triples. If golden triples are below
+        a certain threshold in checksum it's considered golden, otherwise,
+        it's considered a silver triple. This is important for the compression
+        for golden triples because it's used to recover TOF1 because
+        compression scheme to save space on golden triples doesn't send
+        down TOF1 so it's recovered on the ground using the checksum
+    case_number: int
+        The compression case number for the direct event. The case number
+        determines how the bits are arranged in the compressed data.
+    tof_calculation_binary: dict
+        Binary used to determine which TOF coefficients should be used
+        for decompressing the binary.
+    tof_decoder: list
+        Shows how the fields in the binary are split up by bit length and order.
+    remaining_bits: dict
+        The TOF coefficients that should be used for decompression
+    parsed_bits: dict
+        The binary bits split up by TOF field.
+
+    Methods
+    -------
+    __init__(packet, software_vesion, packet_file_name):
+        Constructor, uses the CCSDS packet, version of the software, and
+        the name of the packet file to parse and store information about
+        the Direct Event packet data.
+    decompress_data():
+        Decompresses the Science Direct Event TOF data.
+
+    """
+
+    SHCOARSE: int
+    COUNT: int
+    DATA: bitstring.Bits
+    CHKSUM: int
+    TOF0: float
+    TOF1: float
+    TOF2: float
+    TOF3: float
+    TIME: float
+    ENERGY: float
+    POS: float
+    case_number: int
+    tof_calculation_binary: dict
+    tof_decoder: list
+    remaining_bits: dict
+    parsed_bits: list
+
+    def __init__(self, packet, software_version: str, packet_file_name: str):
+        super().__init__(software_version, packet_file_name, CcsdsData(packet.header))
+        self.parse_data(packet)
+
+    def decompress_data(self):
+        """Decompress the Lo Science Direct Events data."""
+        self._find_decompression_case()
```

Comment:
Might I suggest passing in pieces of `self.DATA` into these methods? eg `self._find_decompression_case(self.DATA[0:4]
self._find_tof_decoder_for_case(self.DATA[4]` 

etc.. it makes it more clear what is dependent on `self.DATA` and also lets you check that you don't have multiple methods acting on the same parts of DATA by accident. That being said, I see later you have a weird dependency for `start_bit` so it maybe wouldn't work so well in that case. 

Also, it's not clear to me, but if any of the class vars aren't expected to be used by external code (for example, if you're only using `self.case_number` to calculate TOF decoders and parsing, but it's meaningless outside of that context) then I'd consider having the method return that value and passing it in to methods that need it. If I see a class variable, I assume it's important for external use cases as well. 

That all being said, it's a nitpick, and I think these methods are clean and testable.

Reply:
I'd agree with @maxinelasp's suggestion, but possibly my preference for plain functions goes even farther where these would just be module-level functions not methods on objects and remove `self` altogether to avoid the potential spurious side-effects you mention.

```python
case = _find_decompression_case(self.DATA[0:4])
decoder = _find_tof_decoder_for_case(self.DATA[4], case)
...
```

Reply:
@maxinelasp Yeah, I was back and forth on that and figured I'd see what others thought. I will pass in and return those values instead.

@greglucas how strong is your preference for that? I definitely have a preference for organizing them as methods, but if that's a standard we want to follow, I will make that change.

Reply:
+0.5, non-blocking, so it is totally fine to do it this way if you prefer.

Just for clarity though, maybe another way to express what I am getting at is to say you have one class to store the initial input, and then another _different_ class to store the decompressed data, so as not to unpack it into the same object structure.

```python
def decompress_data(data: Level0Compressed) -> Level0Decompressed
    # call other functions/methods on the L0 object to get the tables you need
    return new_data_object

decompressed_data_object = decompress_data(input_raw_data)
```

Reply:
Personally, I prefer to keep it organized within the object, because the methods are tightly coupled with the data inside the object and are not very widely applicable outside of the dataclass.